query-graphs/src/postgres.ts

```diff
@@ -281,6 +281,70 @@ function colorForeignScan(node: TreeNode, foreignScan?: string) {
+// Actual Total Time is cumulative
+// Nodes with Actual Loops > 1 record an average Actual Total Time
+// Parallelized nodes have Actual Loops = Workers Launched + 1 for the leader
+// Not all Postgres plans have a root Execution Time even when children have Actual Total Time
+function colorRelativeExecutionTime(root: TreeNode) {
+    let executionTime = root.properties?.get("Execution Time");
+    if (executionTime === undefined) {
+        for (const child of treeDescription.allChildren(root)) {
+            const actualTotalTime = child.properties?.get("Actual Total Time");
+            if (actualTotalTime) {
+                assert(executionTime === undefined, "Unexpected result child node");
+                executionTime = actualTotalTime;
+            }
+        }
+    }
+    if (executionTime) {
+        for (const child of treeDescription.allChildren(root)) {
+            colorChildRelativeExecutionTime(child, Number(executionTime), 1);
+        }
+    }
+}
+function colorChildRelativeExecutionTime(node: TreeNode, executionTime: number, degreeOfParallelism: number) {
+    let childrenTime = 0;
+    if (node.tag === "Gather" || node.tag === "Gather Merge") {
+        const workersLaunched = node.properties?.get("Workers Launched");
+        assert(workersLaunched !== undefined, "Unexpected Workers Launched");
+        degreeOfParallelism = Number(workersLaunched) + 1 /* leader */;
+    }
+    for (const child of treeDescription.allChildren(node)) {
+        const actualTotalTime = child.properties?.get("Actual Total Time");
+        if (actualTotalTime) {
+            const actualLoops = child.properties?.get("Actual Loops");
+            assert(actualLoops !== undefined, "Unexpected Actual Loops");
+            const childLoops = Number(actualLoops) / degreeOfParallelism;
+            childrenTime += Number(actualTotalTime) * childLoops;
+        }
+    }
+    const actualTotalTime = node.properties?.get("Actual Total Time");
+    if (actualTotalTime) {
+        const nodeTotalTime = Number(actualTotalTime);
+        const actualLoops = node.properties?.get("Actual Loops");
+        assert(actualLoops !== undefined, "Unexpected Actual Loops");
+        let nodeLoops = Number(actualLoops);
+        if (node.tag !== "Gather" && node.tag !== "Gather Merge") {
+            nodeLoops /= degreeOfParallelism;
+        }
+
+        const relativeExecutionTime = (nodeTotalTime * nodeLoops - childrenTime) / executionTime;
+        // TODO: remove Actual Total Time of a CTE from referencing CTE Scan subplans
+        // TODO: assert(relativeExecutionTime >= 0, "Unexpected relative execution time");
+
+        assert(node.properties !== undefined);
+        node.properties.set("relativeExecutionTime", relativeExecutionTime.toFixed(3));
+        const l = (100 + (36 - 100) * relativeExecutionTime).toFixed(3);
+        const hsl = "hsl(309, 84%, " + l + "%)";
+        node.properties.set("metricsFill", hsl);
```

Comment:
Putting this into the `properties` map bypasses the type-safety which we tried to reach by using TypeScript.
Instead rather extend the `TreeNode ` class with a new member.
That way, the supported functionality will be better documented/we have a clearer interface between the loaders and the tree-rendering of QueryGraphs

Reply:
Moved fill attributes into TreeNode.
Added fill attributes to debug properties construction.
Kept relative execution time (not used in tree rendering) in properties for the user.
Renamed relative execution time to collate after existing properties without a resort and to locate consistently.